Add new auth flow basics

LukasReschke opened the pull request:
This implements the basics for the new app-password based authentication flow for our clients.
The current implementation tries to keep it as simple as possible and works the following way:

1. Unauthenticated client opens `/index.php/login/flow`
2. User will be asked whether they want to grant access to the client
3. If accepted the user has the chance to do so using existing App Token or automatically generate an app password.

If the user chooses to use an existing app token then that one will simply be redirected to the `nc://` protocol handler.
While we can improve on that in the future, I think keeping this smaller at the moment has its advantages. Also, in the
near future we have to think about an automatic migration endpoint so there's that anyways :-)

If the user chooses to use the regular login the following happens:

1. A session state token is written to the session
2. User is redirected to the login page
3. If successfully authenticated they will be redirected to a page redirecting to the POST controller
4. The POST controller will check if the CSRF token as well as the state token is correct, if yes the user will be redirected to the `nc://` protocol handler.

This approach is quite simple but also allows to be extended in the future. One could for example allow external websites to consume this authentication endpoint as well.


-------

![screen shot 2017-04-24 at 22 44 32](https://cloud.githubusercontent.com/assets/878997/25358974/a8587052-2943-11e7-83da-a9ff625744c2.png)
![screen shot 2017-04-24 at 22 44 30](https://cloud.githubusercontent.com/assets/878997/25358976/a85cae38-2943-11e7-9f69-0d50e459bd13.png)

LukasReschke committed:
Add new auth flow

This implements the basics for the new app-password based authentication flow for our clients.
The current implementation tries to keep it as simple as possible and works the following way:

1. Unauthenticated client opens `/index.php/login/flow`
2. User will be asked whether they want to grant access to the client
3. If accepted the user has the chance to do so using existing App Token or automatically generate an app password.

If the user chooses to use an existing app token then that one will simply be redirected to the `nc://` protocol handler.
While we can improve on that in the future, I think keeping this smaller at the moment has its advantages. Also, in the
near future we have to think about an automatic migration endpoint so there's that anyways :-)

If the user chooses to use the regular login the following happens:

1. A session state token is written to the session
2. User is redirected to the login page
3. If successfully authenticated they will be redirected to a page redirecting to the POST controller
4. The POST controller will check if the CSRF token as well as the state token is correct, if yes the user will be redirected to the `nc://` protocol handler.

This approach is quite simple but also allows to be extended in the future. One could for example allow external websites to consume this authentication endpoint as well.

Signed-off-by: Lukas Reschke <lukas@statuscode.ch>

rullzer commented on core/Controller/ClientFlowLoginController.php:
```diff
@@ -0,0 +1,232 @@
+			],
+			'guest'
+		);
+		$response->setStatus(Http::STATUS_FORBIDDEN);
```
redundant ;)

rullzer commented on core/Controller/ClientFlowLoginController.php:
```diff
@@ -0,0 +1,232 @@
+	private function stateTokenForbiddenResponse() {
+		$response = new TemplateResponse(
+			$this->appName,
+			'403',
```
Http::STATUS_FORBIDDEN

rullzer replied:
never mind :P

rullzer commented on core/Controller/ClientFlowLoginController.php:
```diff
@@ -0,0 +1,232 @@
+		if($this->userSession->isLoggedIn()) {
+			return new TemplateResponse(
+				$this->appName,
+				'403',
```
Http::STATUS_FORBIDDEN

MorrisJobke commented:
@LukasReschke @rullzer Could you explain the background of this?

LukasReschke commented:
> @LukasReschke @rullzer Could you explain the background of this?

Sure. So what we're aiming to do here is make the new authentication approach that we have implemented in that consulting project usable for anybody. Basically, the client can just open a web view, the user has to log in there and gets redirected to the `nc://` protocol redirect. A new app token is then automatically created for that user.

This is required because the push notification code that we have relies on having an app token and won't work with app passwords. Furthermore, this makes using a second factor or another authentication method such as SAML quite easy. 

Also, this gives us the ability to revoke single clients quite easily etc…  – I've added unit tests to this and also added a PR description.

MorrisJobke commented:
> This is required because the push notification code that we have relies on having an app token and won't work with app passwords. Furthermore, this makes using a second factor or another authentication method such as SAML quite easy.

Umm ... what is the difference?

LukasReschke commented:
> Umm ... what is the difference?

/app passwords/regular passwords/. 🙈

MorrisJobke commented:
> /app passwords/regular passwords/. 🙈

Puh - just was shocked shortly, that I used the wrong approach :D

ChristophWurst commented:
- [ ] BUG: redirect URL is lost when I enter a wrong password on the login page (might be an issue of master too, but here it's especially critical)

ChristophWurst commented:
- [ ] BUG: does not work for me on FF.

Steps to reproduce
1. Visit `https://localhost/login/flow`
2. Click "Grant access" -> redirected to `https://localhost/login?redirect_url=/login/flow/redirect%3FstateToken%3Dn3t7CRMbZBlNwdKpZdGOKRc7lyYE2u3ial47hrNjF0z3PkVmdkIh7R6tX7sA8T9P`
3. Log in with correct credentials -> redirected to `https://localhost/login/flow/redirect?stateToken=n3t7CRMbZBlNwdKpZdGOKRc7lyYE2u3ial47hrNjF0z3PkVmdkIh7R6tX7sA8T9P` -> redirected to `https://localhost/login/flow`
4. See `Access forbidden. State token does not match` (HTTP 403)

rullzer committed:
Correctly case the stateToken

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer committed:
Do not remove the state token to early

we should check the stateToken before we remove it. Else the check will
always fail.

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer commented:
>  BUG: does not work for me on FF.

Can you check now?

ChristophWurst commented:
> Can you check now?

Works 👍  Now I'm correctly redirected to `nc://admin:3fmMCHZha2dn%2F7nLo8MnyC2UvtA08J0P33xY251HEEuJTCa0GsjMeoTcZScO@localhost`

Although it might only be shown shortly, this last page looks a bit odd:
![bildschirmfoto von 2017-04-25 10-04-14](https://cloud.githubusercontent.com/assets/1374172/25374684/a5e82122-299e-11e7-803c-8cbdbb046fb3.png)

rullzer commented:
You can work your magic if you want ;)

ChristophWurst reviewed (approved):
Tested and works

ChristophWurst commented:
> You can work your magic if you want ;)

I'm fine with this. Then we're consistent with https://github.com/nextcloud/server/issues/4298 🙊

rullzer committed:
AppToken to 72 chars

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

jancborchardt commented:
Fixed the styling to adjust it to our normal dark instead of light background and proper padding. (Like the `.warning, .update, .error` blocks.)

jancborchardt commented:
I also think I fixed the style of the Redirect page. Please double-check @ChristophWurst @rullzer :)

rullzer commented:
@jancborchardt I don't see your commits. I get some github mess up (I got a lot of unicorns). So could you push your changes again?

jancborchardt committed:
Fix auth flow background color and redirect view layout

Signed-off-by: Jan-Christoph Borchardt <hey@jancborchardt.net>

jancborchardt commented:
Hope it works now and hope it wasn’t a force-push fuckup ;)

rullzer commented:
Looks good @jancborchardt !

rullzer commented on core/routes.php:
```diff
@@ -49,6 +49,9 @@
 		['name' => 'login#confirmPassword', 'url' => '/login/confirm', 'verb' => 'POST'],
 		['name' => 'login#showLoginForm', 'url' => '/login', 'verb' => 'GET'],
 		['name' => 'login#logout', 'url' => '/logout', 'verb' => 'GET'],
+		['name' => 'ClientFlowLogin#showAuthPickerPage', 'url' => '/login/flow', 'verb' => 'GET'],
```
so is `/flow` OK? or do we want something else?

rullzer replied:
As discussed with @LukasReschke lets go with this for now. And at some point come up with anonymous capabilities that would allow auto discovery of the endpoint.

MorrisJobke reviewed (approved):
Works 👍
